addons/controllers/csi/vspherecsiconfig_controller.go

```diff
@@ -0,0 +1,173 @@
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&csiv1alpha1.VSphereCSIConfig{}).
+		WithOptions(options).
+		Complete(r)
+}
+
+func (r *VSphereCSIConfigReconciler) reconcileVSphereCSIConfig(ctx context.Context,
+	csiCfg *csiv1alpha1.VSphereCSIConfig,
+	cluster *clusterapiv1beta1.Cluster) (result ctrl.Result, retErr error) {
+
+	logger := log.FromContext(ctx)
+
+	patchHelper, err := clusterapipatchutil.NewHelper(csiCfg, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		if retErr != nil {
+			// don't modify VSphereCSIConfig if there is an error
+			return
+		}
+
+		if err := patchHelper.Patch(ctx, csiCfg); err != nil {
+			logger.Error(err, "Failed to patch VSphereCSIConfig")
+			retErr = err
+		}
+	}()
+
+	if !csiCfg.GetDeletionTimestamp().IsZero() {
+		return ctrl.Result{}, nil // deleted
+	}
+
+	if result, err = r.reconcileVSphereCSIConfigNormal(ctx, csiCfg, cluster); err != nil {
+		logger.Error(err, "Error reconciling VSphereCSIConfig")
+		return result, err
+	}
+
+	return result, nil
+}
+
+func (r *VSphereCSIConfigReconciler) reconcileVSphereCSIConfigNormal(ctx context.Context,
+	csiCfg *csiv1alpha1.VSphereCSIConfig,
+	cluster *clusterapiv1beta1.Cluster) (ctrl.Result, error) {
+
+	logger := log.FromContext(ctx)
+
+	ownerRef := metav1.OwnerReference{
+		APIVersion: clusterapiv1beta1.GroupVersion.String(),
+		Kind:       cluster.Kind,
+		Name:       cluster.Name,
+		UID:        cluster.UID,
+	}
+
+	if !clusterapiutil.HasOwnerRef(csiCfg.OwnerReferences, ownerRef) {
+		csiCfg.OwnerReferences = clusterapiutil.EnsureOwnerRef(csiCfg.OwnerReferences, ownerRef)
+	}
+
+	addonName := ""
+	switch csiCfg.Spec.VSphereCSI.Mode {
+	case VSphereCSINonParavirtualMode:
+		addonName = constants.CSIAddonName
+	case VSphereCSIParavirtualMode:
+		addonName = constants.PVCSIAddonName
+	}
+
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      util.GenerateDataValueSecretName(cluster.Name, addonName),
+			Namespace: csiCfg.Namespace},
+		Type: v1.SecretTypeOpaque,
+	}
+
+	secret.SetOwnerReferences([]metav1.OwnerReference{ownerRef})
+
+	mutateFn := func() error {
+		secret.Data = make(map[string][]byte)
+		dvs, err := r.mapVSphereCSIConfigToDataValues(ctx, csiCfg, cluster)
+		if err != nil {
+			logger.Error(err, "Error while mapping VSphereCSIConfig to data values")
+			return err
+		}
+		yamlBytes, err := yaml.Marshal(dvs)
+		if err != nil {
+			logger.Error(err, "Error marshaling CSI config data values to yaml")
+			return err
+		}
+		secret.Data[constants.TKGDataValueFileName] = yamlBytes
+		return nil
+	}
+
+	opResult, err := controllerutil.CreateOrPatch(ctx, r.Client, secret, mutateFn)
+
+	if err != nil {
+		logger.Error(err, "Error creating or patching VSphereCSIConfig data values secret")
+		return ctrl.Result{}, err
+	}
+
+	logger.Info(fmt.Sprintf("'%s' the secret '%s'", opResult,
+		fmt.Sprintf("%s/%s", secret.Namespace, secret.Name)))
+
+	csiCfg.Status.SecretRef = &secret.Name
```

Comment:
suggestion: it is risky to mutate the csiCfg pointer directly. I assume you want to do `client.Update()` or `client.Patch` on csiCfg ?